File: port_julia/src/Transport.jl
"""Velocity-space transport integration, port of neort_transport (transport.f90)."""
module Transport

using ..Field
using ..Field: FS
using ..Driftorbit
using ..Driftorbit: DS
using ..Profiles
using ..Profiles: PS
using ..Orbit
using ..Freq
using ..Resonance

export compute_transport_integral

const PI = pi
const SIGN_THETA = -1.0
const OMTH = Ref(0.0)

function d11int(ux, taub, hmn2)
    PI^1.5 * Float64(DS.mph)^2 * Profiles.C^2 * FS.q * PS.vth /
        (PS.qi^2 * DS.dvds * abs(FS.psi_pr)) * ux^3 * exp(-ux^2) * taub * hmn2
end
d12int(ux, taub, hmn2) = d11int(ux, taub, hmn2) * ux^2
function tphi_int(ux, taub, hmn2)
    sgn = FS.psi_pr * FS.q * SIGN_THETA >= 0 ? 1.0 : -1.0
    sgn * PI^1.5 * Float64(DS.mph)^2 * PS.ni1 * Profiles.C * PS.vth / PS.qi *
        ux^3 * exp(-ux^2) * taub * hmn2 * (PS.a1 + PS.a2 * ux^2)
end

function timestep_transport(v, eta, neq, t, y, ydot)
    bmod, _sqrtg, bder, _hcov, hctrvr = do_magfie((get_s(), 0.0, y[1]))
    poloidal_velocity(v, eta, bmod, hctrvr[3], bder[3], y[2], ydot)
    q = FS.q; omth = OMTH[]
    if DS.pertfile
        ar, ai = do_magfie_pert_amp((get_s(), 0.0, y[1]))
        er = DS.epsmn * ar / bmod; ei = DS.epsmn * ai / bmod
    else
        er = DS.epsmn * cos(DS.m0 * y[1]); ei = DS.epsmn * sin(DS.m0 * y[1])
    end
    ph = eta > DS.etatp ? q * DS.mph * y[1] - DS.mth * t * omth :
                          q * DS.mph * y[1] - (DS.mth + q * DS.mph) * t * omth
    cr = cos(ph); ci = sin(ph)
    amp = 2.0 - eta * bmod
    ydot[3] = amp * (er * cr - ei * ci)
    ydot[4] = amp * (er * ci + ei * cr)
    if DS.nonlin
        ydot[5] = 1.0 / bmod; ydot[6] = bmod
    else
        ydot[5] = 0.0; ydot[6] = 0.0
    end
    ydot[7] = 0.0
end

function compute_transport_integral(vmin, vmax, vsteps)
    vth = PS.vth
    roots = zeros(Driftorbit.NLEV * 3)
    d = [0.0, 0.0]; t = 0.0
    du = (vmax - vmin) / (vsteps * vth)
    ux = vmin / vth + du / 2.0
    om_te = PS.om_te; mi = PS.mi
    for _ku in 1:vsteps
        v = ux * vth
        nroots = driftorbit_coarse(v, DS.etamin, DS.etamax, roots, Driftorbit.NLEV)
        for kr in 0:nroots-1
            eta, dres = driftorbit_root(v, 1.0e-8 * abs(om_te), roots[kr*3+1], roots[kr*3+2])
            omth, = om_th(v, eta)
            OMTH[] = omth
            taub = 2.0 * PI / abs(omth)
            bavg, _ = bounce_fast_ext(v, eta, taub, timestep_transport)
            hmn2 = (bavg[3]^2 + bavg[4]^2) * (mi * (ux * vth)^2 / 2.0)^2
            DS.nonlin && error("nonlinear path not ported (outside golden gate)")
            atten = 1.0
            d[1] += du * d11int(ux, taub, hmn2) / abs(dres) * atten
            d[2] += du * d12int(ux, taub, hmn2) / abs(dres) * atten
            if DS.comptorque
                t += du * tphi_int(ux, taub, hmn2) / abs(dres) * atten
            end
        end
        ux += du
    end
    r0 = FS.r0; iota = FS.iota; a = FS.a
    d_plateau = PI * vth^3 / (16.0 * r0 * iota * (PS.qi * DS.b0 / (mi * Profiles.C))^2)
    dsdreff = 2.0 / a * sqrt(get_s())
    d[1] = dsdreff^(-2) * d[1] / d_plateau
    d[2] = dsdreff^(-2) * d[2] / d_plateau
    return d, t
end

end # module

File: port_rust/src/transport.rs
//! Velocity-space transport integration, literal port of neort_transport
//! (transport.f90). Produces D11/D12 and torque density for one harmonic.

use crate::driftorbit as dorb;
use crate::field::{do_magfie, do_magfie_pert_amp, with_field, SIGN_THETA};
use crate::orbit::{bounce_fast_ext, get_s, poloidal_velocity, NVAR};
use crate::profiles;
use crate::freq::om_th;
use crate::resonance::{driftorbit_coarse, driftorbit_root};
use std::cell::Cell;

const PI: f64 = std::f64::consts::PI;

thread_local! {
    static OMTH: Cell<f64> = Cell::new(0.0);
}

fn d11int(ux: f64, taub: f64, hmn2: f64) -> f64 {
    let d = dorb::get();
    let q = with_field(|f| f.q);
    let psi_pr = with_field(|f| f.psi_pr);
    let p = profiles::get();
    PI.powf(1.5) * (d.mph * d.mph) as f64 * profiles::C * profiles::C * q * p.vth
        / (p.qi * p.qi * d.dvds * psi_pr.abs())
        * ux * ux * ux * (-ux * ux).exp() * taub * hmn2
}
fn d12int(ux: f64, taub: f64, hmn2: f64) -> f64 {
    d11int(ux, taub, hmn2) * ux * ux
}
fn tphi_int(ux: f64, taub: f64, hmn2: f64) -> f64 {
    let d = dorb::get();
    let (q, psi_pr) = with_field(|f| (f.q, f.psi_pr));
    let p = profiles::get();
    let sgn = if psi_pr * q * SIGN_THETA >= 0.0 { 1.0 } else { -1.0 };
    sgn * PI.powf(1.5) * (d.mph * d.mph) as f64 * p.ni1 * profiles::C * p.vth / p.qi
        * ux * ux * ux * (-ux * ux).exp() * taub * hmn2 * (p.a1 + p.a2 * ux * ux)
}

pub fn timestep_transport(v: f64, eta: f64, neq: usize, t: f64, y: &[f64], ydot: &mut [f64]) {
    let _ = neq;
    let s = get_s();
    let (bmod, _sqrtg, bder, _hcov, hctrvr) = do_magfie(&[s, 0.0, y[0]]);
    poloidal_velocity(v, eta, bmod, hctrvr[2], bder[2], y[1], ydot);
    let d = dorb::get();
    let q = with_field(|f| f.q);
    let omth = OMTH.with(|c| c.get());

    // epsn (complex)
    let (er, ei) = if d.pertfile {
        let (ar, ai) = do_magfie_pert_amp(&[s, 0.0, y[0]]);
        (d.epsmn * ar / bmod, d.epsmn * ai / bmod)
    } else {
        (d.epsmn * (d.m0 as f64 * y[0]).cos(), d.epsmn * (d.m0 as f64 * y[0]).sin())
    };
    // phase
    let ph = if eta > d.etatp {
        q * d.mph as f64 * y[0] - d.mth as f64 * t * omth
    } else {
        q * d.mph as f64 * y[0] - (d.mth as f64 + q * d.mph as f64) * t * omth
    };
    let (cr, ci) = (ph.cos(), ph.sin());
    let amp = 2.0 - eta * bmod;
    // Hn = amp * epsn * exp(i ph)
    let hr = amp * (er * cr - ei * ci);
    let hi = amp * (er * ci + ei * cr);
    ydot[2] = hr;
    ydot[3] = hi;
    if d.nonlin {
        ydot[4] = 1.0 / bmod;
        ydot[5] = bmod;
    } else {
        ydot[4] = 0.0;
        ydot[5] = 0.0;
    }
    ydot[6] = 0.0;
}

/// Midpoint integral over v: returns (D[2], T).
pub fn compute_transport_integral(vmin: f64, vmax: f64, vsteps: usize) -> ([f64; 2], f64) {
    let p = profiles::get();
    let vth = p.vth;
    let d0 = dorb::get();
    let mut roots = vec![0.0f64; dorb::NLEV * 3];
    let mut d = [0.0f64; 2];
    let mut t = 0.0f64;
    let du = (vmax - vmin) / (vsteps as f64 * vth);
    let mut ux = vmin / vth + du / 2.0;
    let om_te = p.om_te;
    let mi = p.mi;

    for _ku in 1..=vsteps {
        let v = ux * vth;
        let nroots = driftorbit_coarse(v, d0.etamin, d0.etamax, &mut roots, dorb::NLEV);
        for kr in 0..nroots {
            let (eta, dres) = driftorbit_root(v, 1.0e-8 * om_te.abs(), roots[kr * 3], roots[kr * 3 + 1]);
            let (omth, _dv, _de) = om_th(v, eta);
            OMTH.with(|c| c.set(omth));
            let taub = 2.0 * PI / omth.abs();
            let (bavg, _ist): ([f64; NVAR], i32) = bounce_fast_ext(v, eta, taub, timestep_transport);
            let hmn2 = (bavg[2] * bavg[2] + bavg[3] * bavg[3]) * (mi * (ux * vth) * (ux * vth) / 2.0).powi(2);
            let atten = if d0.nonlin {
                panic!("nonlinear attenuation path not ported (outside golden gate)");
            } else {
                1.0
            };
            let dd11 = du * d11int(ux, taub, hmn2) / dres.abs();
            let dd12 = du * d12int(ux, taub, hmn2) / dres.abs();
            d[0] += dd11 * atten;
            d[1] += dd12 * atten;
            if d0.comptorque {
                let dt = du * tphi_int(ux, taub, hmn2) / dres.abs();
                t += dt * atten;
            }
        }
        ux += du;
    }

    let (r0, iota) = with_field(|f| (f.r0, f.iota));
    let a = with_field(|f| f.a);
    let d_plateau = PI * vth.powi(3) / (16.0 * r0 * iota * (p.qi * d0.b0 / (mi * profiles::C)).powi(2));
    let dsdreff = 2.0 / a * get_s().sqrt();
    d[0] = dsdreff.powi(-2) * d[0] / d_plateau;
    d[1] = dsdreff.powi(-2) * d[1] / d_plateau;
    (d, t)
}
